### ISSUE TITLE
[Device Support Request] SIN-4-FP-21 - NodOn

### ISSUE THREAD

**Issue by Mookunicorn** [zigpy/zha-device-handlers#3040]
### Problem description

Missing controls the target temperature of the heater with possible values: comfort, eco, frost_protection, off, comfort_-1, comfort_-2. Currently just an "on, off" switch

### Solution description

Add controls of the heater
(comfort, eco, frost_protection, off, comfort_-1, comfort_-2.)

### Screenshots/Video

<details><summary>Screenshots/Video</summary>

![image](https://github.com/zigpy/zha-device-handlers/assets/6381627/ba5caa17-3027-4931-ab77-8c34ba9a6c34)

</details>

### Device signature

<details><summary>Device signature</summary>

```json
{
  "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.Router: 1>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.FullFunctionDevice|MainsPowered|RxOnWhenIdle|AllocateAddress: 142>, manufacturer_code=4747, maximum_buffer_size=82, maximum_incoming_transfer_size=500, server_mask=11264, maximum_outgoing_transfer_size=500, descriptor_capability_field=<DescriptorCapability.NONE: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, *is_end_device=False, *is_full_function_device=True, *is_mains_powered=True, *is_receiver_on_when_idle=True, *is_router=True, *is_security_capable=False)",
  "endpoints": {
    "1": {
      "profile_id": "0x0104",
      "device_type": "0x0051",
      "input_clusters": [
        "0x0000",
        "0x0003",
        "0x0004",
        "0x0005",
        "0x0006",
        "0x0702",
        "0x1000",
        "0xfc00"
      ],
      "output_clusters": [
        "0x0019"
      ]
    },
    "242": {
      "profile_id": "0xa1e0",
      "device_type": "0x0066",
      "input_clusters": [
        "0x0021"
      ],
      "output_clusters": [
        "0x0021"
      ]
    }
  },
  "manufacturer": "NodOn",
  "model": "SIN-4-FP-21",
  "class": "zigpy.device.Device"
}
```

</details>

### Diagnostic information

<details><summary>Diagnostic information</summary>

```json
{
  "home_assistant": {
    "installation_type": "Home Assistant OS",
    "version": "2024.3.0",
    "dev": false,
    "hassio": true,
    "virtualenv": false,
    "python_version": "3.12.2",
    "docker": true,
    "arch": "x86_64",
    "timezone": "Europe/Paris",
    "os_name": "Linux",
    "os_version": "6.6.20-haos",
    "supervisor": "2024.03.0",
    "host_os": "Home Assistant OS 12.1",
    "docker_version": "24.0.7",
    "chassis": "vm",
    "run_as_root": true
  },
  "custom_components": {
    "zha_toolkit": {
      "version": "v1.1.8",
      "requirements": [
        "pytz"
      ]
    },
    "smartlife": {
      "version": "0.1.0",
      "requirements": [
        "tuya-device-sharing-sdk==0.2.0"
      ]
    },
    "dyson_local": {
      "version": "0.16.4-4",
      "requirements": [
        "libdyson==0.8.11"
      ]
    },
    "espsomfy_rts": {
      "version": "2.4.1",
      "requirements": [
        "websocket-client==1.7.0"
      ]
    },
    "scheduler": {
      "version": "v0.0.0",
      "requirements": []
    },
    "hacs": {
      "version": "1.34.0",
      "requirements": [
        "aiogithubapi>=22.10.1"
      ]
    }
  },
  "integration_manifest": {
    "domain": "zha",
    "name": "Zigbee Home Automation",
    "after_dependencies": [
      "onboarding",
      "usb"
    ],
    "codeowners": [
      "@dmulcahey",
      "@adminiuga",
      "@puddly",
      "@TheJulianJES"
    ],
    "config_flow": true,
    "dependencies": [
      "file_upload"
    ],
    "documentation": "https://www.home-assistant.io/integrations/zha",
    "import_executor": true,
    "iot_class": "local_polling",
    "loggers": [
      "aiosqlite",
      "bellows",
      "crccheck",
      "pure_pcapy3",
      "zhaquirks",
      "zigpy",
      "zigpy_deconz",
      "zigpy_xbee",
      "zigpy_zigate",
      "zigpy_znp",
      "universal_silabs_flasher"
    ],
    "requirements": [
      "bellows==0.38.1",
      "pyserial==3.5",
      "pyserial-asyncio==0.6",
      "zha-quirks==0.0.112",
      "zigpy-deconz==0.23.1",
      "zigpy==0.63.4",
      "zigpy-xbee==0.20.1",
      "zigpy-zigate==0.12.0",
      "zigpy-znp==0.12.1",
      "universal-silabs-flasher==0.0.18",
      "pyserial-asyncio-fast==0.11"
    ],
    "usb": [
      {
        "vid": "10C4",
        "pid": "EA60",
        "description": "*2652*",
        "known_devices": [
          "slae.sh cc2652rb stick"
        ]
      },
      {
        "vid": "10C4",
        "pid": "EA60",
        "description": "*slzb-07*",
        "known_devices": [
          "smlight slzb-07"
        ]
      },
      {
        "vid": "1A86",
        "pid": "55D4",
        "description": "*sonoff*plus*",
        "known_devices": [
          "sonoff zigbee dongle plus v2"
        ]
      },
      {
        "vid": "10C4",
        "pid": "EA60",
        "description": "*sonoff*plus*",
        "known_devices": [
          "sonoff zigbee dongle plus"
        ]
      },
      {
        "vid": "10C4",
        "pid": "EA60",
        "description": "*tubeszb*",
        "known_devices": [
          "TubesZB Coordinator"
        ]
      },
      {
        "vid": "1A86",
        "pid": "7523",
        "description": "*tubeszb*",
        "known_devices": [
          "TubesZB Coordinator"
        ]
      },
      {
        "vid": "1A86",
        "pid": "7523",
        "description": "*zigstar*",
        "known_devices": [
          "ZigStar Coordinators"
        ]
      },
      {
        "vid": "1CF1",
        "pid": "0030",
        "description": "*conbee*",
        "known_devices": [
          "Conbee II"
        ]
      },
      {
        "vid": "0403",
        "pid": "6015",
        "description": "*conbee*",
        "known_devices": [
          "Conbee III"
        ]
      },
      {
        "vid": "10C4",
        "pid": "8A2A",
        "description": "*zigbee*",
        "known_devices": [
          "Nortek HUSBZB-1"
        ]
      },
      {
        "vid": "0403",
        "pid": "6015",
        "description": "*zigate*",
        "known_devices": [
          "ZiGate+"
        ]
      },
      {
        "vid": "10C4",
        "pid": "EA60",
        "description": "*zigate*",
        "known_devices": [
          "ZiGate"
        ]
      },
      {
        "vid": "10C4",
        "pid": "8B34",
        "description": "*bv 2010/10*",
        "known_devices": [
          "Bitron Video AV2010/10"
        ]
      }
    ],
    "zeroconf": [
      {
        "type": "_esphomelib._tcp.local.",
        "name": "tube*"
      },
      {
        "type": "_zigate-zigbee-gateway._tcp.local.",
        "name": "*zigate*"
      },
      {
        "type": "_zigstar_gw._tcp.local.",
        "name": "*zigstar*"
      },
      {
        "type": "_uzg-01._tcp.local.",
        "name": "uzg-01*"
      },
      {
        "type": "_slzb-06._tcp.local.",
        "name": "slzb-06*"
      }
    ],
    "is_built_in": true
  },
  "data": {
    "ieee": "**REDACTED**",
    "nwk": 15611,
    "manufacturer": "NodOn",
    "model": "SIN-4-FP-21",
    "name": "NodOn SIN-4-FP-21",
    "quirk_applied": false,
    "quirk_class": "zigpy.device.Device",
    "quirk_id": null,
    "manufacturer_code": 4747,
    "power_source": "Mains",
    "lqi": 124,
    "rssi": -69,
    "last_seen": "2024-03-13T19:30:27",
    "available": true,
    "device_type": "Router",
    "signature": {
      "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.Router: 1>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.FullFunctionDevice|MainsPowered|RxOnWhenIdle|AllocateAddress: 142>, manufacturer_code=4747, maximum_buffer_size=82, maximum_incoming_transfer_size=500, server_mask=11264, maximum_outgoing_transfer_size=500, descriptor_capability_field=<DescriptorCapability.NONE: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, *is_end_device=False, *is_full_function_device=True, *is_mains_powered=True, *is_receiver_on_when_idle=True, *is_router=True, *is_security_capable=False)",
      "endpoints": {
        "1": {
          "profile_id": "0x0104",
          "device_type": "0x0051",
          "input_clusters": [
            "0x0000",
            "0x0003",
            "0x0004",
            "0x0005",
            "0x0006",
            "0x0702",
            "0x1000",
            "0xfc00"
          ],
          "output_clusters": [
            "0x0019"
          ]
        },
        "242": {
          "profile_id": "0xa1e0",
          "device_type": "0x0066",
          "input_clusters": [
            "0x0021"
          ],
          "output_clusters": [
            "0x0021"
          ]
        }
      },
      "manufacturer": "NodOn",
      "model": "SIN-4-FP-21"
    },
    "active_coordinator": false,
    "entities": [
      {
        "entity_id": "button.chauffage_bureau_identifier",
        "name": "NodOn SIN-4-FP-21"
      },
      {
        "entity_id": "select.chauffage_bureau_comportement_au_demarrage",
        "name": "NodOn SIN-4-FP-21"
      },
      {
        "entity_id": "sensor.nodon_sin_4_fp_21_rssi_3",
        "name": "NodOn SIN-4-FP-21"
      },
      {
        "entity_id": "sensor.nodon_sin_4_fp_21_lqi_3",
        "name": "NodOn SIN-4-FP-21"
      },
      {
        "entity_id": "sensor.chauffage_bureau_demande_instantanee",
        "name": "NodOn SIN-4-FP-21"
      },
      {
        "entity_id": "sensor.chauffage_bureau_summation_delivered",
        "name": "NodOn SIN-4-FP-21"
      },
      {
        "entity_id": "switch.chauffage_bureau_commutateur",
        "name": "NodOn SIN-4-FP-21"
      }
    ],
    "neighbors": [
      {
        "device_type": "Coordinator",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x0000",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "48"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x0934",
        "permit_joining": "Unknown",
        "depth": "15",
        "lqi": "32"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x181E",
        "permit_joining": "Unknown",
        "depth": "15",
        "lqi": "101"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x2D05",
        "permit_joining": "Unknown",
        "depth": "15",
        "lqi": "136"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Parent",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x5DE3",
        "permit_joining": "Unknown",
        "depth": "15",
        "lqi": "163"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0xA6CD",
        "permit_joining": "Unknown",
        "depth": "15",
        "lqi": "95"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0xC414",
        "permit_joining": "Unknown",
        "depth": "15",
        "lqi": "70"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0xD9B0",
        "permit_joining": "Unknown",
        "depth": "15",
        "lqi": "90"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0xEA8F",
        "permit_joining": "Unknown",
        "depth": "15",
        "lqi": "240"
      }
    ],
    "routes": [
      {
        "dest_nwk": "0x0000",
        "route_status": "Active",
        "memory_constrained": false,
        "many_to_one": false,
        "route_record_required": false,
        "next_hop": "0x5DE3"
      }
    ],
    "endpoint_names": [
      {
        "name": "SMART_PLUG"
      },
      {
        "name": "COMBO_BASIC"
      }
    ],
    "user_given_name": "Chauffage Bureau",
    "device_reg_id": "91995077342180562333b3e9aa0ef82e",
    "area_id": "bureau",
    "cluster_details": {
      "1": {
        "device_type": {
          "name": "SMART_PLUG",
          "id": 81
        },
        "profile_id": 260,
        "in_clusters": {
          "0x0000": {
            "endpoint_attribute": "basic",
            "attributes": {
              "0x0001": {
                "attribute_name": "app_version",
                "value": 3
              },
              "0xfffd": {
                "attribute_name": "cluster_revision",
                "value": 3
              },
              "0x0006": {
                "attribute_name": "date_code",
                "value": "2021"
              },
              "0x0014": {
                "attribute_name": "disable_local_config",
                "value": 0
              },
              "0x0003": {
                "attribute_name": "hw_version",
                "value": 52
              },
              "0x0004": {
                "attribute_name": "manufacturer",
                "value": "NodOn"
              },
              "0x0005": {
                "attribute_name": "model",
                "value": "SIN-4-FP-21"
              },
              "0x0007": {
                "attribute_name": "power_source",
                "value": 1
              },
              "0x0002": {
                "attribute_name": "stack_version",
                "value": 103
              },
              "0x4000": {
                "attribute_name": "sw_build_id",
                "value": "3.0.0-1.4.4"
              },
              "0x0000": {
                "attribute_name": "zcl_version",
                "value": 8
              }
            },
            "unsupported_attributes": {}
          },
          "0x0003": {
            "endpoint_attribute": "identify",
            "attributes": {
              "0xfffd": {
                "attribute_name": "cluster_revision",
                "value": 2
              },
              "0x0000": {
                "attribute_name": "identify_time",
                "value": 0
              }
            },
            "unsupported_attributes": {}
          },
          "0x0004": {
            "endpoint_attribute": "groups",
            "attributes": {
              "0xfffd": {
                "attribute_name": "cluster_revision",
                "value": 3
              },
              "0x0000": {
                "attribute_name": "name_support",
                "value": 0
              }
            },
            "unsupported_attributes": {}
          },
          "0x0005": {
            "endpoint_attribute": "scenes",
            "attributes": {
              "0xfffd": {
                "attribute_name": "cluster_revision",
                "value": 3
              },
              "0x0000": {
                "attribute_name": "count",
                "value": 0
              },
              "0x0002": {
                "attribute_name": "current_group",
                "value": 0
              },
              "0x0001": {
                "attribute_name": "current_scene",
                "value": 0
              },
              "0x0004": {
                "attribute_name": "name_support",
                "value": 0
              },
              "0x0003": {
                "attribute_name": "scene_valid",
                "value": 0
              }
            },
            "unsupported_attributes": {}
          },
          "0x0006": {
            "endpoint_attribute": "on_off",
            "attributes": {
              "0xfffd": {
                "attribute_name": "cluster_revision",
                "value": 2
              },
              "0x4000": {
                "attribute_name": "global_scene_control",
                "value": 1
              },
              "0x4002": {
                "attribute_name": "off_wait_time",
                "value": 0
              },
              "0x0000": {
                "attribute_name": "on_off",
                "value": 1
              },
              "0x4001": {
                "attribute_name": "on_time",
                "value": 0
              },
              "0x4003": {
                "attribute_name": "start_up_on_off",
                "value": 255
              }
            },
            "unsupported_attributes": {
              "0x0011": {},
              "0x0001": {}
            }
          },
          "0x0702": {
            "endpoint_attribute": "smartenergy_metering",
            "attributes": {
              "0x0000": {
                "attribute_name": "current_summ_delivered",
                "value": 4738
              },
              "0x000a": {
                "attribute_name": "default_update_period",
                "value": 1
              },
              "0x0304": {
                "attribute_name": "demand_formatting",
                "value": 19
              },
              "0x0302": {
                "attribute_name": "divisor",
                "value": 1000
              },
              "0x0400": {
                "attribute_name": "instantaneous_demand",
                "value": 0
              },
              "0x0306": {
                "attribute_name": "metering_device_type",
                "value": 0
              },
              "0x0301": {
                "attribute_name": "multiplier",
                "value": 1
              },
              "0x0200": {
                "attribute_name": "status",
                "value": 0
              },
              "0x0303": {
                "attribute_name": "summation_formatting",
                "value": 59
              },
              "0x0300": {
                "attribute_name": "unit_of_measure",
                "value": 0
              }
            },
            "unsupported_attributes": {
              "0x0100": {
                "attribute_name": "current_tier1_summ_delivered"
              },
              "0x0102": {
                "attribute_name": "current_tier2_summ_delivered"
              },
              "0x0001": {
                "attribute_name": "current_summ_received"
              },
              "0x0104": {
                "attribute_name": "current_tier3_summ_delivered"
              },
              "0x0106": {
                "attribute_name": "current_tier4_summ_delivered"
              },
              "0x010a": {
                "attribute_name": "current_tier6_summ_delivered"
              },
              "0x0108": {
                "attribute_name": "current_tier5_summ_delivered"
              }
            }
          },
          "0x1000": {
            "endpoint_attribute": "lightlink",
            "attributes": {},
            "unsupported_attributes": {}
          },
          "0xfc00": {
            "endpoint_attribute": "manufacturer_specific",
            "attributes": {},
            "unsupported_attributes": {
              "0x0000": {},
              "0x0001": {},
              "0x0020": {}
            }
          }
        },
        "out_clusters": {
          "0x0019": {
            "endpoint_attribute": "ota",
            "attributes": {
              "0xfffd": {
                "attribute_name": "cluster_revision",
                "value": 4
              },
              "0x0002": {
                "attribute_name": "current_file_version",
                "value": 66564
              }
            },
            "unsupported_attributes": {
              "0x0002": {
                "attribute_name": "current_file_version"
              }
            }
          }
        }
      },
      "242": {
        "device_type": {
          "name": "COMBO_BASIC",
          "id": 102
        },
        "profile_id": 41440,
        "in_clusters": {
          "0x0021": {
            "endpoint_attribute": "green_power",
            "attributes": {},
            "unsupported_attributes": {}
          }
        },
        "out_clusters": {
          "0x0021": {
            "endpoint_attribute": "green_power",
            "attributes": {},
            "unsupported_attributes": {}
          }
        }
      }
    }
  }
}
```

</details>

### Logs

<details><summary>Logs</summary>

```python
[Paste the logs here]
```

</details>

### Custom quirk

<details><summary>Custom quirk (Thanks to Oxillo)</summary>

```python
"""NodOn SIN-4-FP-21 pilot wire module."""
from typing import Any, Final
from zigpy.quirks import CustomCluster
from zigpy.quirks.v2 import add_to_registry_v2, EntityType, EntityPlatform
from zigpy.zcl.foundation import (
    BaseAttributeDefs,
    BaseCommandDefs,
    ZCLAttributeDef,
    ZCLCommandDef,
    WriteAttributesStatusRecord,
    Status
)

from zigpy.zcl import ClusterType

import zigpy.types as t

import logging
_LOGGER = logging.getLogger(__name__)

class NodonPilotWireMode(t.enum16):
    OFF = 0x00
    CONFORT = 0x01
    ECO = 0x02
    HORS_GEL = 0x03
    CONFORT_MOINS_1 = 0x04
    CONFORT_MOINS_2 = 0x05

NODON_PILOT_WIRE_CLUSTER_ID = 0xfc00  #64512

class NodonPilotWireCluster(CustomCluster):
    """NodOn pilot wire manufacturer-specific cluster."""

    cluster_id: Final = NODON_PILOT_WIRE_CLUSTER_ID
    name: Final = "Pilot Wire"
    ep_attribute: Final = "pilot_wire_cluster"

    class AttributeDefs(BaseAttributeDefs):
        pilot_wire_mode: Final = ZCLAttributeDef(id=0x0000, type=NodonPilotWireMode, access="r")

    class ClientCommandDefs(BaseCommandDefs):
        setMode: Final = ZCLCommandDef(
            id=0x0000,
            schema={"mode": NodonPilotWireMode},
            direction=True,
            is_manufacturer_specific=True,
        )
    
    async def write_attributes(
        self, attributes: dict[str | int, Any], manufacturer: int | None = None
    ) -> list:
        """Override writes to the pilot_wire_mode attribute."""
        if "pilot_wire_mode" in attributes:
            await self.client_command(0,attributes.get("pilot_wire_mode"))
            return [[WriteAttributesStatusRecord(Status.SUCCESS)]]
        return await super().write_attributes(attributes, manufacturer=manufacturer)


(
    add_to_registry_v2("NodOn", "SIN-4-FP-21")
    .also_applies_to("ADEO", "SIN-4-FP-21_EQU")
    .replaces(NodonPilotWireCluster)
    .enum("pilot_wire_mode",NodonPilotWireMode,NODON_PILOT_WIRE_CLUSTER_ID,
        ClusterType.Server,1,EntityType.STANDARD,EntityPlatform.SELECT,False,True)
)
```

</details>

**Comment by Skunnyk** [comment]
Related: https://github.com/zigpy/zha-device-handlers/issues/2777

**Comment by Kriss1670** [comment]
Waiting also for ZHA support...

**Comment by oxillo** [comment]
I've just made some progress on it using quirks v2. It is my first quirk and I'm new to HA so I believe there is room for improvement.

Here is the quirk
```python
"""NodOn on/off switch two channels."""
from typing import Final
from zigpy.quirks import CustomCluster
from zigpy.quirks.v2 import add_to_registry_v2, EntityType, EntityPlatform
from zigpy.zcl.foundation import (
    BaseAttributeDefs,
    BaseCommandDefs,
    ZCLAttributeDef,
    ZCLCommandDef
)

from zigpy.zcl import ClusterType

import zigpy.types as t

class NodonPilotWireMode(t.enum16):
    OFF = 0x00
    COMFORT = 0x01
    ECO = 0x02
    FROST_PROTECTION = 0x03
    COMFORT_MINUS_1 = 0x04
    COMFORT_MINUS_2 = 0x05

NODON_PILOT_WIRE_CLUSTER_ID = 0xfc00  #64512

class NodonPilotWireCluster(CustomCluster):
    """NodOn pilot wire manufacturer-specific cluster."""

    cluster_id: Final = NODON_PILOT_WIRE_CLUSTER_ID
    name: Final = "Pilot Wire"
    ep_attribute: Final = "pilot_wire_cluster"
    
    
    class AttributeDefs(BaseAttributeDefs):
        pilot_wire_mode: Final = ZCLAttributeDef(id=0x0000, type=NodonPilotWireMode, access="r")

    class ClientCommandDefs(BaseCommandDefs):
        setMode: Final = ZCLCommandDef(
            id=0x0000,
            schema={
                "mode": NodonPilotWireMode
            },
            direction=True,
            is_manufacturer_specific=True,
        )


(
    add_to_registry_v2("NodOn", "SIN-4-FP-21")
    .replaces(NodonPilotWireCluster)
    .enum("pilot_wire_mode",NodonPilotWireMode,NODON_PILOT_WIRE_CLUSTER_ID,
          ClusterType.Server,1,EntityType.STANDARD,EntityPlatform.SELECT,False,True,"pilot_wire_mode")
)
```

And the result in entity view
![Nodon](https://github.com/zigpy/zha-device-handlers/assets/13654859/94d2f0e1-00da-4ad6-bc2a-f031d6995686)

**Comment by oxillo** [comment]
And with an addition `also_applies_to()` that should solve #2777 

````python
(
    add_to_registry_v2("NodOn", "SIN-4-FP-21")
    .also_applies_to("ADEO", "SIN-4-FP-21_EQU")
    .replaces(NodonPilotWireCluster)
    .enum("pilot_wire_mode",NodonPilotWireMode,NODON_PILOT_WIRE_CLUSTER_ID,
          ClusterType.Server,1,EntityType.STANDARD,EntityPlatform.SELECT,False,True,"pilot_wire_mode")
)
```

**Comment by Mookunicorn** [comment]
Thanks for your work, the Quirks applies correctly but when I try to change modes there is an error message that says
`Échec de l'appel du service select/select_option. Failed to write attribute pilot_wire_mode=<NodonPilotWireMode.ECO: 2> :<Status.INVALID_DATA_TYPE: 141>`

Maybe I did something wrong

<details><summary>Custom quirk</summary>

```python
"""NodOn pilot wire switch two channels."""
from typing import Final
from zigpy.quirks import CustomCluster
from zigpy.quirks.v2 import add_to_registry_v2, EntityType, EntityPlatform
from zigpy.zcl.foundation import (
    BaseAttributeDefs,
    BaseCommandDefs,
    ZCLAttributeDef,
    ZCLCommandDef
)

from zigpy.zcl import ClusterType

import zigpy.types as t

class NodonPilotWireMode(t.enum16):
    OFF = 0x00
    COMFORT = 0x01
    ECO = 0x02
    FROST_PROTECTION = 0x03
    COMFORT_MINUS_1 = 0x04
    COMFORT_MINUS_2 = 0x05

NODON_PILOT_WIRE_CLUSTER_ID = 0xfc00  #64512

class NodonPilotWireCluster(CustomCluster):
    """NodOn pilot wire manufacturer-specific cluster."""

    cluster_id: Final = NODON_PILOT_WIRE_CLUSTER_ID
    name: Final = "Pilot Wire"
    ep_attribute: Final = "pilot_wire_cluster"
    
    
    class AttributeDefs(BaseAttributeDefs):
        pilot_wire_mode: Final = ZCLAttributeDef(id=0x0000, type=NodonPilotWireMode, access="r")

    class ClientCommandDefs(BaseCommandDefs):
        setMode: Final = ZCLCommandDef(
            id=0x0000,
            schema={
                "mode": NodonPilotWireMode
            },
            direction=True,
            is_manufacturer_specific=True,
        )

(
    add_to_registry_v2("NodOn", "SIN-4-FP-21")
    .also_applies_to("ADEO", "SIN-4-FP-21_EQU")
    .replaces(NodonPilotWireCluster)
    .enum("pilot_wire_mode",NodonPilotWireMode,NODON_PILOT_WIRE_CLUSTER_ID,ClusterType.Server,1,EntityType.STANDARD,EntityPlatform.SELECT,False,True,"pilot_wire_mode")
)
```
</details>

The problem occurs on all 5 modules.


<details><summary>Error logs</summary>

```
Enregistreur: homeassistant.components.websocket_api.http.connection
Source: components/websocket_api/commands.py:239
intégration: Home Assistant WebSocket API ([documentation](https://www.home-assistant.io/integrations/websocket_api), [problèmes](https://github.com/home-assistant/core/issues?q=is%3Aissue+is%3Aopen+label%3A%22integration%3A+websocket_api%22))
S'est produit pour la première fois: 14:23:47 (1 occurrences)
Dernier enregistrement: 14:23:47

[139849444259520] Unexpected exception
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 239, in handle_call_service
    response = await hass.services.async_call(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2738, in async_call
    response_data = await coro
                    ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2779, in _execute_service
    return await target(service_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 975, in entity_service_call
    single_response = await _handle_entity_call(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 1047, in _handle_entity_call
    result = await task
             ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 191, in async_handle_select_option
    await self.async_select_option(option)
  File "/usr/src/homeassistant/homeassistant/components/zha/select.py", line 227, in async_select_option
    await self._cluster_handler.write_attributes_safe(
  File "/usr/src/homeassistant/homeassistant/components/zha/core/cluster_handlers/__init__.py", line 552, in write_attributes_safe
    raise HomeAssistantError(
homeassistant.exceptions.HomeAssistantError: Failed to write attribute pilot_wire_mode=<NodonPilotWireMode.COMFORT_MINUS_1: 4>: <Status.INVALID_DATA_TYPE: 141>
```
</details>

**Comment by oxillo** [comment]
Hello, 
I'm using the quirk on the latest version of HA (2024.5.3). What version do you use ? 
Quirks v2 are new and support is evolving at every version.

**Comment by Mookunicorn** [comment]
I'm on 2024.5.3 too 🤔
![Screenshot_20240514-210825.png](https://github.com/zigpy/zha-device-handlers/assets/6381627/b5d2e4c3-3378-4f29-87a1-5482dd136c82)

**Comment by Mookunicorn** [comment]
I wonder if the problem could be with the ZigBee key.

I use SkyConnect without multiprotocol

**Comment by oxillo** [comment]
This is not your installation; it is my solution. 
The enum entity tries to write the `pilot_wire_mode` attribute which is not supported. You must use the command to change the mode.

**Comment by oxillo** [comment]
Some update to catch attribute write and replace by commands 

```python
"""NodOn SIN-4-FP-21 pilot wire module."""
from typing import Any, Final
from zigpy.quirks import CustomCluster
from zigpy.quirks.v2 import add_to_registry_v2, EntityType, EntityPlatform
from zigpy.zcl.foundation import (
    BaseAttributeDefs,
    BaseCommandDefs,
    ZCLAttributeDef,
    ZCLCommandDef,
    WriteAttributesStatusRecord,
    Status
)

from zigpy.zcl import ClusterType

import zigpy.types as t

import logging
_LOGGER = logging.getLogger(__name__)

class NodonPilotWireMode(t.enum16):
    OFF = 0x00
    COMFORT = 0x01
    ECO = 0x02
    FROST_PROTECTION = 0x03
    COMFORT_MINUS_1 = 0x04
    COMFORT_MINUS_2 = 0x05

NODON_PILOT_WIRE_CLUSTER_ID = 0xfc00  #64512

class NodonPilotWireCluster(CustomCluster):
    """NodOn pilot wire manufacturer-specific cluster."""

    cluster_id: Final = NODON_PILOT_WIRE_CLUSTER_ID
    name: Final = "Pilot Wire"
    ep_attribute: Final = "pilot_wire_cluster"

    class AttributeDefs(BaseAttributeDefs):
        pilot_wire_mode: Final = ZCLAttributeDef(id=0x0000, type=NodonPilotWireMode, access="r")

    class ClientCommandDefs(BaseCommandDefs):
        setMode: Final = ZCLCommandDef(
            id=0x0000,
            schema={"mode": NodonPilotWireMode},
            direction=True,
            is_manufacturer_specific=True,
        )
    
    async def write_attributes(
        self, attributes: dict[str | int, Any], manufacturer: int | None = None
    ) -> list:
        """Override writes to the pilot_wire_mode attribute."""
        if "pilot_wire_mode" in attributes:
            await self.client_command(0,attributes.get("pilot_wire_mode"))
            return [[WriteAttributesStatusRecord(Status.SUCCESS)]]
        return await super().write_attributes(attributes, manufacturer=manufacturer)


(
    add_to_registry_v2("NodOn", "SIN-4-FP-21")
    .also_applies_to("ADEO", "SIN-4-FP-21_EQU")
    .replaces(NodonPilotWireCluster)
    .enum("pilot_wire_mode",NodonPilotWireMode,NODON_PILOT_WIRE_CLUSTER_ID,
        ClusterType.Server,1,EntityType.STANDARD,EntityPlatform.SELECT,False,True)
)
```

**Comment by Mookunicorn** [comment]
Oh my boooy it work well.
I just took the liberty of translating the different modes. Do you want me to add the Quirk in my initial request (of course mentioning you are the boss)

**Comment by oxillo** [comment]
Happy to see it works for you !
You can add the quirk to your initial request.
I would like to work on the HMI but I have other projects in parallel that requires my attention. I'll look into it in some weeks.

**Comment by RenZan** [comment]
Hello, 
I used the quirk but I have often this error and all zigbee integration fails to load since I used it:

```
Enregistreur: homeassistant.config_entries
Source: config_entries.py:594
S'est produit pour la première fois: 20:54:16 (1 occurrences)
Dernier enregistrement: 20:54:16

Error setting up entry Multiprotocol add-on - Nabu Casa for zha
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 594, in async_setup
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/zha/__init__.py", line 134, in async_setup_entry
    async with radio_mgr.connect_zigpy_app() as app:
  File "/usr/local/lib/python3.12/contextlib.py", line 210, in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/zha/radio_manager.py", line 182, in connect_zigpy_app
    app = await self.radio_type.controller.new(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/zigpy/application.py", line 252, in new
    await app._load_db()
  File "/usr/local/lib/python3.12/site-packages/zigpy/application.py", line 108, in _load_db
    await self._dblistener.load()
  File "/usr/local/lib/python3.12/site-packages/zigpy/appdb.py", line 675, in load
    device = zigpy.quirks.get_device(device)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/zigpy/quirks/__init__.py", line 41, in get_device
    return _DEVICE_REGISTRY.get_device(device)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/zigpy/quirks/registry.py", line 96, in get_device
    raise MultipleQuirksMatchException(
zigpy.exceptions.MultipleQuirksMatchException: Multiple matches found for device <Device model='SIN-4-FP-21' manuf='NodOn' nwk=0x409D ieee=bc:02:6e:ff:fe:ae:2b:b4 is_initialized=True>: [QuirksV2RegistryEntry(registry=<zigpy.quirks.registry.DeviceRegistry object at 0x7f4937aea120>, quirk_location='file[/config/custom_components/zha_custom/quirks/nodon_pilot_wire.py]-line:56', filters=[], custom_device_class=None, device_node_descriptor=None, skip_device_configuration=False, adds_metadata=[], removes_metadata=[], replaces_metadata=[ReplacesMetadata(remove=RemovesMetadata(cluster_id=64512, endpoint_id=1, cluster_type=<ClusterType.Server: 0>), add=AddsMetadata(cluster=<class 'nodon_pilot_wire.NodonPilotWireCluster'>, endpoint_id=1, cluster_type=<ClusterType.Server: 0>, constant_attributes={}))], entity_metadata=[ZCLEnumMetadata(entity_platform=<EntityPlatform.SELECT: 'select'>, entity_type=<EntityType.STANDARD: 'standard'>, cluster_id=64512, endpoint_id=1, cluster_type=<ClusterType.Server: 0>, initially_disabled=False, attribute_initialized_from_cache=True, translation_key=None, enum=<enum 'NodonPilotWireMode'>, attribute_name='pilot_wire_mode')], device_automation_triggers_metadata={}), QuirksV2RegistryEntry(registry=<zigpy.quirks.registry.DeviceRegistry object at 0x7f4937aea120>, quirk_location='file[/config/custom_components/zha_custom/quirks/nodon_pilot_wire.py]-line:56', filters=[], custom_device_class=None, device_node_descriptor=None, skip_device_configuration=False, adds_metadata=[], removes_metadata=[], replaces_metadata=[ReplacesMetadata(remove=RemovesMetadata(cluster_id=64512, endpoint_id=1, cluster_type=<ClusterType.Server: 0>), add=AddsMetadata(cluster=<class 'nodon_pilot_wire.NodonPilotWireCluster'>, endpoint_id=1, cluster_type=<ClusterType.Server: 0>, constant_attributes={}))], entity_metadata=[ZCLEnumMetadata(entity_platform=<EntityPlatform.SELECT: 'select'>, entity_type=<EntityType.STANDARD: 'standard'>, cluster_id=64512, endpoint_id=1, cluster_type=<ClusterType.Server: 0>, initially_disabled=False, attribute_initialized_from_cache=True, translation_key=None, enum=<enum 'NodonPilotWireMode'>, attribute_name='pilot_wire_mode')], device_automation_triggers_metadata={})]

```
any idea how to avoid this problem ?

thank you.

**Comment by rungeard** [comment]
> any idea how to avoid this problem ?

@RenZan it seems to be related to https://github.com/zigpy/zigpy/issues/1410

**Comment by anoril1979** [comment]
Hi!
And update from @garc33
https://github.com/zigpy/zha-device-handlers/issues/2777#issuecomment-2358979497

**Comment by ikruglov** [comment]
Hi.

Without realizing this issue, I prepared a PR to add a [quirk for NodON SIN-4-FP-21](https://github.com/zigpy/zha-device-handlers/pull/3364). It contains quirk implementation and tests. It works in my setup. I have installed a NodON SIN-4-FP-21 to control my bathroom's electric floor heating module.

This issue has working code snippets, but I didn't find ready PRs. Please correct me if I'm wrong.
What blocks the issue from being merged into the codebase?

P.S. I guess my code is missing the `QuirkBuilder` part, which is simple to fix.

**Comment by oxillo** [comment]
Thanks @ikruglov for initiating a PR. 
I did not initiated it by myself as I'm quite new to HA and thought that my additions were not mature enough.
Indeed, I would like that the pilot wire mode entity to have a name as "select.radiateur_sam _pilotwiremode" and not as "select.radiateur_sam_none" as of today. i would appreciate any help of this.
I would also like to have a default card for pilot wire.

For QuirkBuilder, it is the new way to define quirks (v2). The PR #3364 uses the old style; so it should not be needed.

**Comment by ikruglov** [comment]
Thanks for the clarification @oxillo. Let's join forces to get it well-prepared and merged.

I have a question. Why did you need special handing of `write_attributes()`?  In my case, I just defined an attribute and a command. As a result, both setting and getting pilot_wire modes work without any extra handling.

Also, I've tried using the v2 interface with QuirkBuilder, but tests have started failing. I haven't found a way to keep them working so far. So, I'll probably keep the code in its current state unless the code owner objects.

**Comment by ikruglov** [comment]
@oxillo I've re-created my PR. It now uses `QuirkBuilder()`. I decided to use it because it has a much cleaner interface and very concise code. I still lack tests, but I've asked for maintainer help with an example.

The code works in my case. I can get and set pilot wire mode in both the HomeAsistant interface and directly manipulating cluster via the 'Manage Zigbee device.' It required no extra code for me or specific handling.

I also understood what you mean by "radiateur_sam_none." Yeah, I also didn't find a way. But it's a simple problem. The entity can be renamed in the HA interface, so I'll leave it for now.

**Comment by jyrivallan** [comment]
@ikruglov may I ask you to take into account https://github.com/zigpy/zha-device-handlers/issues/2777 at the same time? `SIN-4-FP-21_EQU` and `SIN-4-FP-21` are exactly the same device.
I'm no expert, but it seems it could be as simple as a `.also_applies_to("NodOn","SIN-4-FP-21")` in the QuirkBuilder

**Comment by ikruglov** [comment]
@jyrivallan sure thing. I've added it to the PR. But I have no way of testing it. So, you will need to do it.

**Comment by oxillo** [comment]
@ikruglov : About the write_attributes() , it is a trick to be able to have a combobox to control the pilot wire mode in the ZHA entity page (see picture in #3040). 
We could also have just the mode value displayed and 6 buttons to set each individual mode (one for OFF, one for COMFORT, one for ECO,...)

**Comment by garc33** [comment]
> @jyrivallan sure thing. I've added it to the PR. But I have no way of testing it. So, you will need to do it.

I comment in the PR it won't work. The adeo device needs to override the manufacturer code.

**Comment by ikruglov** [comment]
> @ikruglov : About the write_attributes() , it is a trick to be able to have a combobox to control the pilot wire mode in the ZHA entity page (see picture in #3040). We could also have just the mode value displayed and 6 buttons to set each individual mode (one for OFF, one for COMFORT, one for ECO,...)

@oxillo thanks for the clarification. With my PR I'm getting the same result I think.  And no extra handling.

<img width="673" alt="Screenshot 2024-09-22 at 22 14 31" src="https://github.com/user-attachments/assets/43cf882c-7fd4-4abe-aec1-ed4dc941e19c">

**Comment by ikruglov** [comment]
> > @jyrivallan sure thing. I've added it to the PR. But I have no way of testing it. So, you will need to do it.
> 
> I comment in the PR it won't work. The adeo device needs to override the manufacturer code.

@garc33 I don't see your comment. Where did you leave it?

**Comment by garc33** [comment]
> > > @jyrivallan sure thing. I've added it to the PR. But I have no way of testing it. So, you will need to do it.
> 
> > 
> 
> > I comment in the PR it won't work. The adeo device needs to override the manufacturer code.
> 
> 
> 
> @garc33 I don't see your comment. Where did you leave it?

https://github.com/zigpy/zha-device-handlers/pull/3364#discussion_r1770588689

**Comment by ikruglov** [comment]
> > > > @jyrivallan sure thing. I've added it to the PR. But I have no way of testing it. So, you will need to do it.
> > 
> > 
> > > 
> > 
> > 
> > > I comment in the PR it won't work. The adeo device needs to override the manufacturer code.
> > 
> > 
> > @garc33 I don't see your comment. Where did you leave it?
> 
> [#3364 (comment)](https://github.com/zigpy/zha-device-handlers/pull/3364#discussion_r1770588689)

For some reason, I don't see this comment.

If a one-line change doesn't work, I will revert it since I cannot test changes myself. Someone with the ADEO device is welcome to follow up on my PR.

**Comment by anoril1979** [comment]
Hello!
I tried this quirk.
The device attribute names changed, so I guess the quirk is applied, but I don't have any selector for the pilot wire mode.

**Comment by anoril1979** [comment]
Hello!
I'm struggling with Home Assistant for this one!
I added the quirk and restarted HA. Then I unpaired my devices and refresh. For a moment, they were there, displaying the pilotwire mode and I thought it was ok. But changing the mode did nothing on the device.
I refreshed and restarted and went into "multipledeviceexception" on ZHA side. I realized my devices were all named "SIN-4..." I removed them and restarted. Impossible to get ride of them.
Well, in the end, I refreshed, restarted and now I have my 3 devices back.
But they are considered as "zigpy.device.Device" and the quirk is not dealing with them.
I'm a bit lost.
Note: the device work fine: I can commute and have power consumption feedback. But no wire mode.
![image](https://github.com/user-attachments/assets/f48fb9de-bbff-49bb-a884-259ae2180fb7)
![image](https://github.com/user-attachments/assets/613f1636-43a2-40bf-b498-6d1bfd6eba78)
![image](https://github.com/user-attachments/assets/84cea8dc-151f-41fc-a4ab-ef875c4d3352)

**Comment by anoril1979** [comment]
I removed all my Adeo devices. I restarted HA by rebooting the rapsberry.
Hmmm, my devices are in pairing mode (by themself, weird) and "search new devices" does not seems to find them anymore.

**Comment by Mookunicorn** [comment]
Since 2024.11.0 HA update, quirk doesn't work.

From what I understand from the logs, the attribute "'fallback_name'" is missing in ".enum". I therefore tried adding it as follows without success:
```python
.enum("pilot_wire_mode", NodonPilotWireMode, NODON_PILOT_WIRE_CLUSTER_ID,
    ClusterType.Server, 1, EntityType.STANDARD, EntityPlatform.SELECT, False, True, fallback_name="mode")
```

<details><summary>Error Log</summary>
Enregistreur: zhaquirks
Source: /usr/local/lib/python3.12/site-packages/zhaquirks/__init__.py:481
S'est produit pour la première fois: 10:49:36 (1 occurrences)
Dernier enregistrement: 10:49:36

Unexpected exception importing custom quirk 'NodOnPiloteWire'
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/zhaquirks/__init__.py", line 479, in setup
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 995, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/config/custom_zha_quirks/NodOnPiloteWire.py", line 78, in <module>
    .enum(
     ^^^^^
  File "/usr/local/lib/python3.12/site-packages/zigpy/quirks/v2/__init__.py", line 666, in enum
    ZCLEnumMetadata(
  File "<attrs generated init zigpy.quirks.v2.ZCLEnumMetadata>", line 16, in __init__
    __attr_validator_fallback_name(self, __attr_fallback_name, self.fallback_name)
  File "/usr/local/lib/python3.12/site-packages/attr/validators.py", line 101, in __call__
    raise TypeError(
TypeError: ("'fallback_name' must be <class 'str'> (got None that is a <class 'NoneType'>).", Attribute(name='fallback_name', default=NOTHING, validator=<instance_of validator for type <class 'str'>>, repr=True, eq=True, eq_key=None, order=True, order_key=None, hash=None, init=True, metadata=mappingproxy({}), type='str', converter=None, kw_only=True, inherited=True, on_setattr=None, alias='fallback_name'), <class 'str'>, None)
</details>
<details><summary>Custom quirk</summary>

```python
"""Nodon SIN-4-FP-21 | Adeo SIN-4-FP-21_EQU"""
from typing import Any, Final
from zigpy.quirks import CustomCluster
from zigpy.quirks.v2 import QuirkBuilder, EntityType, EntityPlatform
from zigpy.zcl.foundation import (
    BaseAttributeDefs,
    BaseCommandDefs,
    ZCLAttributeDef,
    ZCLCommandDef,
    WriteAttributesStatusRecord,
    Status
)

from zigpy.zcl import ClusterType

import zigpy.types as t

import logging
_LOGGER = logging.getLogger(__name__)

class NodonPilotWireMode(t.enum16):
    OFF = 0x00
    COMFORT = 0x01
    ECO = 0x02
    FROST_PROTECTION = 0x03
    COMFORT_MINUS_1 = 0x04
    COMFORT_MINUS_2 = 0x05

NODON_PILOT_WIRE_CLUSTER_ID = 0xfc00  #64512

class NodonPilotWireCluster(CustomCluster):
    """NodOn pilot wire manufacturer-specific cluster."""

    cluster_id: Final = NODON_PILOT_WIRE_CLUSTER_ID
    name: Final = "Pilot Wire"
    ep_attribute: Final = "pilot_wire_cluster"
    manufacturer_id_override: Final = 4747
    
    class AttributeDefs(BaseAttributeDefs):
        pilot_wire_mode: Final = ZCLAttributeDef(
            id=0x0000, 
            type=NodonPilotWireMode, 
            access="r"
        )

    class ClientCommandDefs(BaseCommandDefs):
        setMode: Final = ZCLCommandDef(
            id=0x0000,
            schema={"mode": NodonPilotWireMode},
            direction=True,
            is_manufacturer_specific=True
        )
    
    async def write_attributes(
        self, attributes: dict[str | int, Any], manufacturer: int | None = None
    ) -> list:
        """Override writes to the pilot_wire_mode attribute."""
        if "pilot_wire_mode" in attributes:
            await self.client_command(0, attributes.get("pilot_wire_mode"), manufacturer=4747)
            await self.read_attributes(attributes)
            return [[WriteAttributesStatusRecord(Status.SUCCESS)]]
        return await super().write_attributes(attributes, manufacturer=4747)

    async def read_attributes(
        self,
        attributes: list[int | str],
        allow_cache: bool = False,
        only_cache: bool = False,
        manufacturer: int | t.uint16_t | None = None,
    ) -> Any:
        return await super().read_attributes(attributes, allow_cache=False, only_cache=False, manufacturer=4747)

(

    QuirkBuilder("ADEO", "SIN-4-FP-21_EQU")
    .also_applies_to("NodOn","SIN-4-FP-21")
    .replaces(NodonPilotWireCluster)
    .enum("pilot_wire_mode",NodonPilotWireMode,NODON_PILOT_WIRE_CLUSTER_ID,
        ClusterType.Server,1,EntityType.STANDARD,EntityPlatform.SELECT,False,True,)

    .add_to_registry()
)
```
</details>

**Comment by ikruglov** [comment]
@Mookunicorn check https://github.com/zigpy/zha-device-handlers/pull/3364. I've pushed a fix for 2024.11.0. It works in my setup

**Comment by ikruglov** [comment]
FYI: https://github.com/zigpy/zha-device-handlers/pull/3364 got merged. It supports both NodOn and Adeo. The code should become part of the 2024.12 release. Unfortunately, the change comes with only ZHA attributes (one which changes via "Manage Zigbee Device") and not with enum in the HA interface due to concerns from maintainers.

**Comment by maximelefrancois86** [comment]
OK so if we want this, for now, we need to remove it from the device registry and add our own quirk.

This works with versions

Core 2024.12.5
Supervisor 2024.12.0
Operating System 14.1
User interface 20241127.8

```python
"""Adeo SIN-4-FP-21_EQU pilot wire module."""
from typing import Any, Final
from zigpy.quirks import CustomCluster, DEVICE_REGISTRY
from zigpy.quirks.v2 import QuirkBuilder, EntityType, EntityPlatform
from zigpy.zcl.foundation import (
    BaseAttributeDefs,
    BaseCommandDefs,
    ZCLAttributeDef,
    ZCLCommandDef,
    WriteAttributesStatusRecord,
    Status
)

from zigpy.zcl import ClusterType

import zigpy.types as t

import logging
_LOGGER = logging.getLogger(__name__)

del DEVICE_REGISTRY._registry_v2[("NodOn","SIN-4-FP-21")]

class NodonPilotWireMode(t.enum16):
    off = 0x00
    confort = 0x01
    eco = 0x02
    hors_gel = 0x03
    confort_moins_1 = 0x04
    confort_moins_2 = 0x05

NODON_PILOT_WIRE_CLUSTER_ID = 0xfc00  #64512

class NodonPilotWireCluster(CustomCluster):
    """NodOn pilot wire manufacturer-specific cluster."""

    cluster_id: Final = NODON_PILOT_WIRE_CLUSTER_ID
    name: Final = "Pilot Wire"
    ep_attribute: Final = "pilot_wire_cluster"
    manufacturer_id_override: Final = 4747
    
    class AttributeDefs(BaseAttributeDefs):
        pilot_wire_mode: Final = ZCLAttributeDef(
            id=0x0000, 
            type=NodonPilotWireMode, 
            access="r"
        )

    class ClientCommandDefs(BaseCommandDefs):
        setMode: Final = ZCLCommandDef(
            id=0x0000,
            schema={"mode": NodonPilotWireMode},
            direction=True,
            is_manufacturer_specific=True
        )
    
    async def write_attributes(
        self, attributes: dict[str | int, Any], manufacturer: int | None = None
    ) -> list:
        """Override writes to the pilot_wire_mode attribute."""
        if "pilot_wire_mode" in attributes:
            await self.client_command(0, attributes.get("pilot_wire_mode"), manufacturer=NodonPilotWireCluster.manufacturer_id_override)
            await self.read_attributes(attributes)
            return [[WriteAttributesStatusRecord(Status.SUCCESS)]]
        return await super().write_attributes(attributes, manufacturer=NodonPilotWireCluster.manufacturer_id_override)

    async def read_attributes(
        self,
        attributes: list[int | str],
        allow_cache: bool = False,
        only_cache: bool = False,
        manufacturer: int | t.uint16_t | None = None,
    ) -> Any:
        return await super().read_attributes(attributes, allow_cache=False, only_cache=False, manufacturer=NodonPilotWireCluster.manufacturer_id_override)

(

    QuirkBuilder("NodOn","SIN-4-FP-21")
    .replaces(NodonPilotWireCluster)
    .enum(
        attribute_name="pilot_wire_mode",
        enum_class=NodonPilotWireMode,
        cluster_id=NODON_PILOT_WIRE_CLUSTER_ID,
        cluster_type=ClusterType.Server,
        endpoint_id=1,
        entity_platform=EntityPlatform.SELECT,
        entity_type=EntityType.STANDARD,
        initially_disabled=False,
        attribute_initialized_from_cache=True,
        reporting_config=None,
        translation_key="mode_fil_pilote",
        fallback_name="mode fil pilote")
    .add_to_registry()
)

```


and for completeness, below is an example of an action in some automation that changes the select

```yaml
- action: select.select_option
      target:
        entity_id:
        - select.chauffage_theo_mode_fil_pilote
        - select.chauffage_noe_mode_fil_pilote
      data:
        option: confort moins 1
```
